### ISSUE TITLE
0.12-rc1 - import writes out empty fields into the state file as `null` instead of `[]` or `{}`, which results in a large diff requiring resource replacement

### ISSUE THREAD

**Issue by a13xb** [hashicorp/terraform#21266]
I don't know if it's provider-specific issue, but my gut feeling it could be more general, since it is related to how Terraform 0.12 writes array attributes into state files, it doesn't happen with 0.11.

### Summary

When you import an EC2 instance (in a VPC) with 0.12, some fields are set to `null`, specifically: `ebs_block_device`, `ephemeral_block_device`, `ipv6_addresses`, `network_interface`, `security_groups`, `volume_tags`.

These fields are set to either `[]` or `{}` for new instances created with Terraform, so it's specific to import.

The problem is when you try plan or apply from this state, Terraform insists that these attributes must be changed, some forcing resource replacement (`ipv6_addresses`, `security_groups`, `ephemeral_block_device`).

If I edit the state file manually and set these attributes to `[]` or `{}` (for volume tags), the diff goes away completely.

### Terraform Version
```
Terraform v0.12.0-rc1
+ provider.aws v1.60.0-dev20190216h00-dev
```

### Terraform Configuration Files
This is what my configuration looked like, but it doesn't seem to matter.

```hcl
resource aws_instance my_instance {
  availability_zone = "ap-southeast-2a"
  ami = "ami-0dad20bd1b9c8c004"
  associate_public_ip_address = true
  key_name = "key-name"
  instance_type = "m3.medium"
  iam_instance_profile = "my-profile"
  monitoring = true
  vpc_security_group_ids = [
    "sg-dd4dcf1e7d0102b5c",
    "sg-ae0386db141b8750f"
  ]
  subnet_id = "subnet-109e955a7a3ae95f0"
  tags = {
    Name = "instance-name"
  }
  lifecycle {
    ignore_changes = ["user_data"]
  }
  root_block_device {
    volume_type = "gp2"
    volume_size = 10
  }
}
```

### Expected Behavior
After running `terraform import` for a resource, running `terraform plan` should result in no changes.

### Actual Behavior
Terraform insists on replacing the instance.

### Steps to Reproduce
1. Create an EC2 instance outside of Terraform
2. Import it
3. Run `terraform plan`

**Comment by a13xb** [comment]
Still seeing this in 0.12.0.

I've tested with more resources since then, and In addition to the above mentioned, I also see it in `aws_vpc_peering_connection` (when `requester` and `accepter` optional blocks are omitted), `aws_route53_zone` (when optional `vpc` block is omitted).

**Comment by a13xb** [comment]
Additionally, Terraform-side-only properties `aws_vpc_peering_connection` `auto_accept` or `aws_route53_zone` `force_destroy` are also set to null.

If with an imported `aws_route53_zone` resource, there will always be a diff regardless of what `force_destroy` is set to.
```
      + force_destroy = false
```

**Comment by pselle** [comment]
Thanks for checking on this @a13xb! I'm going to reopen this to ensure we get to look at it again.

**Comment by alexng-canuck** [comment]
We are also seeing this issue in the OCI provider (but affecting map attributes). See #21737. Setting an empty map results in "null" being written to state file.

**Comment by jbardin** [comment]
Hi @a13xb,

These were actually a couple separate issues showing up here, though they appear at the same time. I tried a couple resources here with the latest core and SDK versions with no diff after import. Can you verify if any of your examples still exhibit this behavior?

**Comment by jbardin** [comment]
While I no longer saw the behavior with the resources and configurations I tested, there are still cases where block values could cause diffs in some situations, which should be addressed in #21764.

**Comment by a13xb** [comment]
Got around to checking it again, I do not see this anymore with v0.12.3.
